spacy/cli/apply.py

```diff
@@ -0,0 +1,133 @@
+
+
+def _stream_jsonl(path: Path) -> Iterable[str]:
+    """
+    Stream "text" field from JSONL. If the field "text" is
+    not found it raises error.
+    """
+    for entry in srsly.read_jsonl(path):
+        if "text" not in entry:
+            raise ValueError(
+                "JSONL files have to contain 'text' field."
+            )
+        else:
+            yield entry["text"]
+
+
+def _maybe_read_text(path: Path) -> Union[str, None]:
+    """
+    Try to read the text file from the provided path.
+    When encoutering a decoding error just warn and pass.
+    """
+    with open(path, 'r') as fin:
+        try:
+            text = fin.read()
+            return text
+        except UnicodeDecodeError as e:
+            msg.warn(f"Skipping file {path}")
+            return None
+
+
+def _stream_texts(paths: List[Path]) -> Iterable[Union[str, None]]:
+    """
+    Yields strings or None when decoding error is encountered.
+    """
+    for path in paths:
+        yield _maybe_read_text(path)
+
+
+@app.command("apply")
+def apply_cli(
+    # fmt: off
+    model: str = Arg(..., help="Model name or path"),
+    data_path: Path = Arg(..., help=path_help, exists=True),
+    output: Path = Arg(..., help=out_help, dir_okay=False),
+    code_path: Optional[Path] = Opt(None, "--code", "-c", help=code_help),
+    use_gpu: int = Opt(-1, "--gpu-id", "-g", help="GPU ID or -1 for CPU."),
+    batch_size: int = Opt(1, "--batch-size", "-b", help="Batch size."),
+    n_process: int = Opt(1, "--n-process", "-n", help="number of processors to use.")
+):
+    """
+    Apply a trained pipeline to documents to get predictions.
+    Expects a loadable spaCy pipeline and some data as input.
+    The data can be provided multiple formats. It can be a single
+    .spacy file or a single text file with one document per line.
+    A directory can also be provided in which case the 'suffix'
+    argument is required. All paths pointing to files with the
+    provided suffix will be recursively collected and processed.
+
+    DOCS: https://spacy.io/api/cli#tba
+    """
+    import_code(code_path)
+    setup_gpu(use_gpu)
+    apply(data_path, output, model, batch_size, n_process)
+
+
+def apply(
+    data_path: Path,
+    output: Path,
+    model: str,
+    batch_size: int,
+    n_process: int,
+):
+    data_path = ensure_path(data_path)
+    output_path = ensure_path(output)
+    if not data_path.exists():
+        msg.fail("Couldn't find data path.", data_path, exits=1)
+    nlp = load_model(model)
+    msg.good(f"Loaded model {model}")
+    vocab = nlp.vocab
+    docbin = DocBin()
+    paths = walk_directory(data_path)
+    streams: List[Union[Iterable[str], Iterable[Doc]]] = []
+    text_files = []
+    for path in paths:
+        if path.suffix == ".spacy":
+            streams.append(_stream_docbin(path, vocab))
+        elif path.suffix == ".jsonl":
+            streams.append(_stream_jsonl(path))
+        else:
+            text_files.append(path)
```

Comment:
I was expecting to see `_stream_text` implemented in the same way as the other two. Is there a reason for the difference?

Reply:
The other two stream from a single file, but the `_stream_text` takes a list of paths as input. The
`for path in paths` loop create the `_stream_docbin` and `_stream_jsonl` generators and collects the rest of the paths into the `text_files` list. 

Then if we found any text files the generator gets created:

```
if len(text_files) > 0:
        streams.append(filter(None, _stream_texts(text_files)))
```

Its also different from the other in that sometimes it yields `None` if when reading some files we encounter `UnicodeDecodeError`. That's why its also wrapped in `filter(None, ...)`.

Reply:
It's not that I don't see how this implementation works in general, but that I think it would be simpler and easier to maintain if it were implemented the same way as `_stream_jsonl`, where the method would just handle a single file and yield a single text from that file if it can open it, and deal with potential warnings/errors in the same way. So that we don't have one method that raises an error and one that warns and returns `None`s that have be filtered out.

Reply:
I didn't know what you mean by it not being "the same". I'll change it to raise error instead of return None.

Reply:
The current version still warns in one place and raises an error in another similar place. For this functionality, I think I would lean toward warning, but if you think errors make more sense that is also fine.

I really want to see a `_stream_text` function that works the exact same way as `_stream_jsonl` for a file with one line instead of this mixture of behaviors and warnings/errors.

Reply:
ok they both `raise` now